### PR TITLE
draft: extra openpgp features

### PR DESCRIPTION
**Pull request by dagheyman**

### CODE REVIEW

**Comment by dagheyman** [comment]
This pull request **introduces 1 alert** when merging 5140051f4b12ac579dd38720016745f15da97114 into 6eff7fadbafb1b0f565a968d0be3de037de9ec3d - [view on LGTM.com](https://lgtm.com/projects/g/Yubico/yubikey-manager/rev/pr-04d90b187083a751de525ca5d996145e898d93c2)

**new alerts:**

* 1 for Unused import

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by dagheyman** [comment]
This pull request **introduces 1 alert** when merging ac78cf673f8f61de8b22a9094ab16399ecc160f0 into 6eff7fadbafb1b0f565a968d0be3de037de9ec3d - [view on LGTM.com](https://lgtm.com/projects/g/Yubico/yubikey-manager/rev/pr-040a097195db4b2103aefc5b81eee52d1b0b69a3)

**new alerts:**

* 1 for Unused import

---

This project has automated code review enabled, but doesn't use the <a href="https://github.com/apps/lgtm-com">LGTM GitHub App</a>. Migrate over by <a href="https://github.com/apps/lgtm-com/installations/new">installing the app</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: ykman/opgp.py [hunk 1]
```diff
@@ -127,23 +208,141 @@ def _verify(self, pw, pin):
...
+        curve = key.curve.name
+        if curve == 'secp384r1':
+            return '2B81040022'
+        if curve == 'secp256r1':
+            return '2A8648CE3D030107'
+        if curve == 'secp521r1':
+            return '2B81040023'
+        if curve == 'x25519':
+            return '2B060104019755010501'
+        raise ValueError('No OID for curve!')
```
Comment (emlun) [inline ykman/opgp.py]:
```suggestion
        raise ValueError('No OID for curve: ' + curve)
```

File: ykman/opgp.py [hunk 2]
```diff
@@ -127,23 +208,141 @@ def _verify(self, pw, pin):
...
+
+    def _get_opgp_algo_id_from_ec(self, key):
+        curve = key.curve.name
+        if curve in ['secp384r1', 'secp256r1', 'secp521r1']:
+            return 0x13
+        if curve == 'x25519':
+            return 0x16
+        raise ValueError('No Algo ID for curve!')
```
Comment (emlun) [inline ykman/opgp.py]:
```suggestion
        raise ValueError('No Algo ID for curve: ' + curve)
```

File: ykman/cli/opgp.py [hunk 3]
```diff
@@ -153,34 +150,42 @@ def echo_default_pins():
...
+def set_touch(ctx, key, policy, admin_pin, force):
     """
-    Manage touch policy for OpenPGP keys.
+    Set touch policy for OpenPGP keys.
 
     \b
-    KEY     Key slot to set (sig, enc or aut).
-    POLICY  Touch policy to set (on, off or fixed).
+    KEY     Key slot to set (sig, enc, aut or att).
+    POLICY  Touch policy to set (on, off, fixed, cached or cached-fix).
```
[the remark this answers is not in the document]
Comment (dagheyman) [reply ykman/cli/opgp.py]: Typo indeed, thanks.

File: ykman/cli/opgp.py [hunk 4]
```diff
@@ -116,17 +105,25 @@ def info(ctx):
...
-        'Signature key           {.name}'.format(
-            controller.get_touch(KEY_SLOT.SIGNATURE)))
-    click.echo(
-        'Encryption key          {.name}'.format(
-            controller.get_touch(KEY_SLOT.ENCRYPTION)))
-    click.echo(
-        'Authentication key      {.name}'.format(
-            controller.get_touch(KEY_SLOT.AUTHENTICATION)))
+    # Touch only available on YK4 and later
+    if controller.version >= (4, 2, 6):
```
[the remark this answers is not in the document]
Comment (dagheyman) [reply ykman/cli/opgp.py]: Done! The bool version feels a bit much though.

File: ykman/opgp.py [hunk 5]
```diff
@@ -118,32 +201,163 @@ def reset(self):
...
 
-    def set_touch(self, key_slot, mode, pin):
-        if self.version < (4, 2, 0):
+    def set_touch(self, key_slot, mode, admin_pin):
+        if not self.supported_touch_policies:
             raise ValueError('Touch policy is available on YubiKey 4 or later.')
-        self._verify(PW3, pin)
-        self.send_apdu(0, INS.PUT_DATA, 0, key_slot,
+        if mode not in self.supported_touch_policies:
+            raise ValueError('Touch policy not available on this device.')
```
Comment (emlun) [inline ykman/opgp.py]: Can we communicate which touch policy is not working, maybe something like this?

```suggestion
            raise ValueError('Touch policy {} requested, device supports only: {}'.format(mode, self.supported_touch_policies))
```

**Pull request merged**